components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java

```diff
@@ -447,6 +448,11 @@ public void preGetRoleIdListOfUser(String userId, String tenantDomain) throws Id
     public void postGetRoleIdListOfUser(List<String> roleIds, String userId, String tenantDomain)
             throws IdentityRoleManagementException {
 
+        if (!isFragmentApplication()) {
+            return;
+        }
+
+        filterAndRemoveNonSharedRoles(roleIds, tenantDomain);
```

Comment:
I think it will be better to implement this logic in `getAssociatedRolesOfApplication` method in `ApplicationManagementService` which is used by in `getRolesAssociatedWithApplication ` method in `AppAssociatedRolesResolverImpl.java`. Relying on thread local properties in service level will not be an ideal approach since we can't guarantee setting thread local properties in every place that this service is used.